### PR TITLE
Refine Metadata Module API

### PR DESCRIPTION
**Pull request by dwedul-figure**
## Description

Change some endpoint and field names in the Metadata module API to hopefully make things clearer and easier for others to use.

closes: #169

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by dwedul-figure** [comment]
The numbers make the diff look more daunting than it is. Most of the 19k additions and 7k removals are auto-generated pieces.
```
      Repo Directory: provenance
         From Branch: main
           To Branch: dwedul/169-metadata-refinements
========================================
Line Changes -  Code:  +2125   -1128
Line Changes - Proto:   +853    -294
Line Changes -  Auto: +15199   -5256
Line Changes - Tests:   +871    -614
Line Changes - Total: +19048   -7292
```

File: proto/provenance/metadata/v1/query.proto [hunk 1]
```diff
@@ -21,96 +21,198 @@ service Query {
...
-  // Scope returns a specific scope by id
+  // ---- Primary Data Queries -----
+
+  // Scope searches for a scope.
+  //
+  // The scope id, if provided, must either be scope uuid, e.g. 91978ba2-5f35-459a-86a7-feca1b0512e0 or a scope address,
+  // e.g. scope1qzge0zaztu65tx5x5llv5xc9ztsqxlkwel. The session addr, if provided, must be a bech32 session address,
+  // e.g. session1qxge0zaztu65tx5x5llv5xc9zts9sqlch3sxwn44j50jzgt8rshvqyfrjcr. The record_addr, if provided, must be a
+  // bech32 record address, e.g. record1q2ge0zaztu65tx5x5llv5xc9ztsw42dq2jdvmdazuwzcaddhh8gmu3mcze3. Case doesn't
+  // matter.
```
Comment (iramiller) [inline proto/provenance/metadata/v1/query.proto:32]: Case is defined by the specifications of bech32 encoding to be all lower case, so in that sense it does matter even if we don't require it to match thei requirement.

File: proto/provenance/metadata/v1/query.proto [hunk 2]
```diff
@@ -21,96 +21,198 @@ service Query {
...
-  rpc SessionContextByID(SessionContextByIDRequest) returns (SessionContextByIDResponse) {
-    option (google.api.http).get = "/provenance/metadata/v1/scope/id/{scope_id}/session/{session_id}";
+  // Sessions searches for sessions.
+  //
+  // The scope id can either be scope uuid, e.g. 91978ba2-5f35-459a-86a7-feca1b0512e0 or a scope address, e.g.
+  // scope1qzge0zaztu65tx5x5llv5xc9ztsqxlkwel. Similarly, the session_id can either be a uuid or session address, e.g.
+  // session1qxge0zaztu65tx5x5llv5xc9zts9sqlch3sxwn44j50jzgt8rshvqyfrjcr. Case doesn't matter.
+  //
+  // * If only a session id is provided, and it is an address, then the scope id will be ignored and the session id used
+  // to find a single session.
```
Comment (iramiller) [inline proto/provenance/metadata/v1/query.proto:66]: If the scope id within the session id doesn't match the included scope-id this seems more like an error condition (invalid request/id mis-match) than an ignore case.

**Review by iramiller — COMMENTED** [review]
This is a bit of a monster.. some of the refactoring (such as removing the must deserialize from keeper state store reads feels like it should have been a separate PR if only to reduce this overall size to something a bit more manageable to process.  Schedule is tight and objectives are high so I understand the desire to push in larger sets of changes but in the future note that multi-issue refactor PRs are likely to be denied/request to be split up if for no other reason than they can't be given a fair assessment otherwise.

The inclusion of the original request in all of the responses to queries is another area of concern.  What is the reasoning here? For an RPC model a request is linked to a response in the callers context.

All of that said some of the work in making the CLI usable is particularly brilliant and well thought out.  I expect that the flexible parameters there will be a huge boon to users trying to understand and query the metadata module contents during development.

File: proto/provenance/metadata/v1/query.proto [hunk 3]
```diff
@@ -119,239 +221,460 @@ service Query {
...
-  string           scope_uuid = 4 [(gogoproto.moretags) = "yaml:\"scope_uuid\""];
+  // scope is the wrapped scope result.
+  ScopeWrapper scope = 1;
+  // sessions is any number of wrapped sessions in this scope (if requested).
+  repeated SessionWrapper sessions = 2 [(gogoproto.moretags) = "yaml:\"sessions,omitempty\""];
+  // records is any number of wrapped records in this scope (if requested).
+  repeated RecordWrapper records = 3 [(gogoproto.moretags) = "yaml:\"records,omitempty\""];
+
+  // request is a copy of the request that generated these results.
+  ScopeRequest request = 98;
```
[the remark this answers is not in the document]
Comment (dwedul-figure) [reply proto/provenance/metadata/v1/query.proto:260]: I figured that it'd allow for a fire and forget pattern if someone wanted it. But I also figured it'd be useful for troubleshooting by users. If they're not getting the response they expected, maybe the request will contain clues.

File: proto/provenance/metadata/v1/query.proto [hunk 4]
```diff
@@ -119,239 +221,460 @@ service Query {
...
-  cosmos.base.query.v1beta1.PageRequest pagination = 1;
-}
-
 // QueryParamsResponse is the response type for the Query/Params RPC method.
 message QueryParamsResponse {
   // params defines the parameters of the module.
   Params params = 1 [(gogoproto.nullable) = false];
+
+  // request is a copy of the request that generated these results.
+  QueryParamsRequest request = 98;
```
[the remark this answers is not in the document]
Comment (dwedul-figure) [reply proto/provenance/metadata/v1/query.proto:230]: Just keeping it in-line with the rest of the responses.

File: proto/provenance/metadata/v1/query.proto [hunk 5]
```diff
@@ -21,96 +21,198 @@ service Query {
...
 
-  // Scope returns a specific scope by id
+  // ---- Primary Data Queries -----
+
+  // Scope searches for a scope.
+  //
+  // The scope id, if provided, must either be scope uuid, e.g. 91978ba2-5f35-459a-86a7-feca1b0512e0 or a scope address,
+  // e.g. scope1qzge0zaztu65tx5x5llv5xc9ztsqxlkwel. The session addr, if provided, must be a bech32 session address,
+  // e.g. session1qxge0zaztu65tx5x5llv5xc9zts9sqlch3sxwn44j50jzgt8rshvqyfrjcr. The record_addr, if provided, must be a
+  // bech32 record address, e.g. record1q2ge0zaztu65tx5x5llv5xc9ztsw42dq2jdvmdazuwzcaddhh8gmu3mcze3. Case doesn't
```
[the remark this answers is not in the document]
Comment (dwedul-figure) [reply proto/provenance/metadata/v1/query.proto:31]: Oh shoot. You're right. Case doesn't matter for the UUIDs. But we don't lower-case it in `MetadataAddressFromBech32`. So I'll remove that part of the documentation here (and a couple other places I put it).

File: proto/provenance/metadata/v1/query.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (dwedul-figure) [reply proto/provenance/metadata/v1/query.proto:66]: I was on the fence on that while writing this. I'll switch it to an error instead of ignoring it.

**Review by iramiller — APPROVED** [review]
Awesome.

**Pull request merged**